scapy/contrib/isotp.py

```diff
@@ -57,289 +54,1346 @@ def answers(self, other):
-                           extended_rx_addr=None,
-                           listen_only=False):
-        option_flags = CAN_ISOTP_DEFAULT_FLAGS
-        if extended_addr is not None:
-            option_flags = option_flags | CAN_ISOTP_EXTEND_ADDR
-        else:
-            extended_addr = CAN_ISOTP_DEFAULT_EXT_ADDRESS
+    def validate_fields(self):
+        if self.src is not None:
+            if type(self.src) is not int or not 0 <= self.src < 0x20000000:
+                raise Exception("src is not a valid CAN identifier")
+        if self.dst is not None:
+            if type(self.dst) is not int or not 0 <= self.dst < 0x20000000:
+                raise Exception("dst is not a valid CAN identifier")
+        if self.exsrc is not None:
+            if type(self.exsrc) is not int or not 0 <= self.exsrc < 0x100:
+                raise Exception("exsrc is not a byte")
+        if self.exdst is not None:
+            if type(self.exdst) is not int or not 0 <= self.exdst < 0x100:
+                raise Exception("exdst is not a byte")
 
-        if extended_rx_addr is not None:
-            option_flags = option_flags | CAN_ISOTP_RX_EXT_ADDR
-        else:
-            extended_rx_addr = CAN_ISOTP_DEFAULT_EXT_ADDRESS
+    def fragment(self):
+        data_bytes_in_frame = 7
+        if self.exdst is not None:
+            data_bytes_in_frame = 6
+
+        if len(self.data) >= 4096:
+            raise Exception("Too much data in ISOTP message")
+
+        if len(self.data) <= data_bytes_in_frame:
+            # We can do this in a single frame
+            frame_data = struct.pack('B', len(self.data)) + self.data
+            if self.exdst:
+                frame_data = struct.pack('B', self.exdst) + frame_data
+            pkt = CAN(identifier=self.dst, data=frame_data)
+            return [pkt]
+
+        idx = 0
+        n = 0
+        pkts = []
+        while idx < len(self.data):
+            frame_data = self.data[idx:idx + data_bytes_in_frame]
+            if n == 0:
+                # first frame (one less data byte)
+                frame_data = frame_data[:-1]
+                frame_header = struct.pack(">H", len(self.data) + 0x1000)
+            else:
+                frame_header = struct.pack("b", (n % 16) + 0x20)
+
+            n += 1
+            idx += len(frame_data)
+
+            if self.exdst:
+                frame_header = struct.pack('B', self.exdst) + frame_header
+            pkt = CAN(identifier=self.dst, data=frame_header + frame_data)
+            pkts.append(pkt)
+        return pkts
+
+    @staticmethod
+    def defragment(can_frames, use_extended_addressing=None):
+        assert (len(can_frames) > 0)
+
+        dst = can_frames[0].identifier
+        for frame in can_frames:
+            if frame.identifier != dst:
+                raise Exception("Not all CAN frames have the same identifier")
+            if len(frame.data) <= 0:
+                raise Exception("Some CAN frames are empty")
```

Comment:
I would prefer to use a warning rather than an exception
`scapy.error.warning("text")`